asmcomp/cmm.mli

```diff
@@ -99,24 +114,38 @@ type memory_chunk =
   | Sixteen_signed
   | Thirtytwo_unsigned
   | Thirtytwo_signed
-  | Word_int                           (* integer or pointer outside heap *)
-  | Word_val                           (* pointer inside heap or encoded int *)
+  | Word of gc_action
   | Single
   | Double                             (* 64-bit-aligned 64-bit float *)
   | Double_u                           (* word-aligned 64-bit float *)
 
-and operation =
+(** Types of pointers to statically-allocated code and data. *)
+type symbol_kind =
+  | Function
+  (** A pointer to the code of a function. *)
+  | Value
+  (** A pointer to a block outside the OCaml heap, which does not (itself and
+      transitively) point back into the heap, which has the correct structure
+      for an OCaml value.  Such blocks should be marked black to avoid wasting
+      GC time.  (They must be marked black if in a read-only section.) *)
+  | Other
+  (** A pointer to arbitrary out-of-heap data. *)
+
+val machtype_component_of_symbol_kind : symbol_kind -> machtype_component
+(** The register type required to hold the address of a symbol with the
+    given kind. *)
+
+type operation =
     Capply of machtype
   | Cextcall of string * machtype * bool * label option
   | Cload of memory_chunk * Asttypes.mutable_flag
   | Calloc
   | Cstore of memory_chunk * Lambda.initialization_or_assignment
-  | Caddi | Csubi | Cmuli | Cmulhi | Cdivi | Cmodi
+  | Cadd of gc_action  (* the [gc_action] gives the type of the result *)
+  | Csubi | Cmuli | Cmulhi | Cdivi | Cmodi
   | Cand | Cor | Cxor | Clsl | Clsr | Casr
```

Comment:
We've just realised that more (all?) of the arithmetic operations will need altering so that they specify the type of the thing (e.g. tagged integer, naked int32, etc...) that they are working on.